source/fundamentals/crud/write-operations/pkFactory.txt

```diff
@@ -0,0 +1,77 @@
+.. _node-pkfactory:
+
+=================================
+Generate Custom Vales for ``_id``
+=================================
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn how to use the MongoDB Node.js driver to generate your
+own ``_id`` values using the **primary key factory**.
+
+The primary key factory allows you to create unique identifiers in your
+documents when you choose not to specify an ``_id`` during an
+:ref:`insert operation <node-fundamentals-insert-data>`. The 
+default primary key factory generates ``ObjectId`` values.
+
+.. note:: Upsert Operations
+
+   The driver doesn't use the primary key factory for
+   :ref:`upsert operations <node-fundamentals-upsert>`. Instead, the
+   driver always creates ``ObjectId`` values for the ``_id`` field of
+   the upserted documents.
+
+   If you want to use the primary key factory, peform a :ref:`find
+   operation <node-fundamentals-retrieve-data>`, then an :ref:`update
+   <node-fundamentals-change-a-document>` or :ref:`insert
+   <node-fundamentals-insert-data>` operation.
+
+Specify a Primay Key Factory
+----------------------------
+
+To specify a primary key factory, apply the ``pkFactory`` option to your
+``MongoClient`` instance.
+
+The following code snippet indicates the ``pkFactory`` option to
+generate ``_id`` values of type ``uuid``:
+
+.. code-block:: javascript
+
+   const { UUID } = require('bson');
+   ...
+   const client = new MongoClient(uri, { 
+     useUnifiedTopology: true, 
+     pkFactory: { createPk: () =>  new UUID().toBinary() }
+   });
+
+.. note:: TypeScript Data Consistency
```

Comment:
From looking at the deleted section of `typescript.txt`, I think the actual point we'd want to discuss here is the following:

If a user is using a `pkFactory` and the user attempts to insert a document that already has an `_id` field **and** that `_id` field is of a different type than the type returned by the `pkFactory`, then they will have inconsistent data.  

This behavior is a result of how the driver applies `pkFactory` The driver only applies `pkFactory` if the document provided by the user does not have an `_id` property.  So something like the following generates two documents with different `_id` types:

```typescript 
const client = new MongoClient(uri, {
   // pkFactory returns a number
    pkFactory: { createPk: () => Math.floor(Math.random() * Number.MAX_SAFE_INTEGER) }, 
  });

....
// inserts document with a string id
collection.insertOne({ _id: "hello", name: "Scar" });  

 // pkFactory will be used to generate a number id
collection.insertOne({ name: "Ursula" });         
```

This behavior exists in both Javascript and Typescript